File: src/main/java/com/github/theonlytails/ruby/registries/BlockRegistry.java
package com.github.theonlytails.ruby.registries;

import com.github.theonlytails.ruby.RubyMod;
import com.github.theonlytails.ruby.blocks.CentrifugeBlock;
import com.github.theonlytails.ruby.blocks.RubyBarrelBlock;
import com.github.theonlytails.ruby.blocks.RubyCarpetBlock;
import com.github.theonlytails.ruby.blocks.RubyOreBlock;
import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.block.material.MaterialColor;
import net.minecraftforge.common.ToolType;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;

public class BlockRegistry {
    public static final DeferredRegister<Block> BLOCKS = DeferredRegister.create(ForgeRegistries.BLOCKS, RubyMod.MOD_ID);

    public static final RegistryObject<Block> RUBY_BLOCK = BLOCKS.register(
            "ruby_block",
            () -> new Block(Block.Properties.create(Material.IRON)
                    .hardnessAndResistance(5.0f, 6.0f)
                    .sound(SoundType.METAL)
                    .harvestTool(ToolType.PICKAXE)
                    .harvestLevel(2)
                    .setRequiresTool())
    );

    public static final RegistryObject<Block> RUBY_SLAB = BLOCKS.register(
            "ruby_slab",
            () -> new SlabBlock(AbstractBlock.Properties.from(RUBY_BLOCK.get()))
    );

    public static final RegistryObject<Block> RUBY_STAIRS = BLOCKS.register(
            "ruby_stairs",
            () -> new StairsBlock(() -> RUBY_BLOCK.get().getDefaultState(), AbstractBlock.Properties.from(RUBY_BLOCK.get()))
    );

    public static final RegistryObject<OreBlock> RUBY_ORE_BLOCK = BLOCKS.register("ruby_ore", RubyOreBlock::new);

    public static final RegistryObject<Block> CENTRIFUGE_BLOCK = BLOCKS.register("centrifuge", CentrifugeBlock::new);

    public static final RegistryObject<Block> RUBY_WOOL = BLOCKS.register(
            "ruby_wool",
            () -> new Block(Block.Properties
                    .create(Material.WOOL, MaterialColor.CRIMSON_HYPHAE)
                    .hardnessAndResistance(0.8f)
                    .sound(SoundType.CLOTH))
    );

    public static final RegistryObject<Block> RUBY_CARPET = BLOCKS.register("ruby_carpet", RubyCarpetBlock::new);

    public static final RegistryObject<Block> RUBY_BARREL = BLOCKS.register("ruby_barrel", RubyBarrelBlock::new);
}

File: src/main/kotlin/com/github/theonlytails/rubymod/registries/BlockRegistry.kt
package com.github.theonlytails.rubymod.registries

import com.github.theonlytails.rubymod.RubyMod
import com.github.theonlytails.rubymod.blocks.CentrifugeBlock
import com.github.theonlytails.rubymod.blocks.RubyBarrelBlock
import com.github.theonlytails.rubymod.blocks.RubyCarpetBlock
import net.minecraft.block.*
import net.minecraft.block.PressurePlateBlock.Sensitivity
import net.minecraft.block.material.Material
import net.minecraft.block.material.MaterialColor
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IWorldReader
import net.minecraftforge.common.ToolType
import net.minecraftforge.registries.ForgeRegistries
import thedarkcolour.kotlinforforge.forge.KDeferredRegister

object BlockRegistry {
	val BLOCKS = KDeferredRegister(ForgeRegistries.BLOCKS, RubyMod.MOD_ID)

	val RUBY_BLOCK by BLOCKS.register("ruby_block") {
		Block(AbstractBlock.Properties.create(Material.IRON)
			.hardnessAndResistance(5.0f, 6.0f)
			.sound(SoundType.METAL)
			.harvestTool(ToolType.PICKAXE)
			.harvestLevel(2)
			.setRequiresTool())
	}

	val RUBY_SLAB by BLOCKS.register("ruby_slab") {
		SlabBlock(AbstractBlock.Properties.from(RUBY_BLOCK))
	}

	val RUBY_PRESSURE_PLATE by BLOCKS.register("ruby_pressure_plate") {
		PressurePlateBlock(Sensitivity.MOBS, AbstractBlock.Properties.from(RUBY_BLOCK))
	}

	val RUBY_BUTTON by BLOCKS.register("ruby_button") {
		StoneButtonBlock(AbstractBlock.Properties.from(RUBY_BLOCK))
	}

	val RUBY_STAIRS by BLOCKS.register("ruby_stairs") {
		StairsBlock({ RUBY_BLOCK.defaultState },
			AbstractBlock.Properties.from(RUBY_BLOCK))
	}

	val RUBY_ORE_BLOCK by BLOCKS.register("ruby_ore") {
		object : OreBlock(Properties.create(Material.ROCK)
			.hardnessAndResistance(3.0f, 3.0f)
			.sound(SoundType.STONE)
			.harvestTool(ToolType.PICKAXE)
			.harvestLevel(2)
			.setRequiresTool()) {
			override fun getExpDrop(
				state: BlockState,
				reader: IWorldReader,
				pos: BlockPos,
				fortune: Int,
				silktouch: Int,
			) = 3
		}
	}

	val CENTRIFUGE_BLOCK by BLOCKS.register("centrifuge", ::CentrifugeBlock)

	val RUBY_WOOL by BLOCKS.register("ruby_wool") {
		Block(AbstractBlock.Properties
			.create(Material.WOOL, MaterialColor.CRIMSON_HYPHAE)
			.hardnessAndResistance(0.8f)
			.sound(SoundType.CLOTH))
	}

	val RUBY_CARPET by BLOCKS.register("ruby_carpet", ::RubyCarpetBlock)

	val RUBY_BARREL by BLOCKS.register("ruby_barrel", ::RubyBarrelBlock)
}